blocks/log/log.py

```diff
@@ -0,0 +1,136 @@
+"""The event-based main loop of Blocks."""
+import sqlite3
+from abc import ABCMeta
+from collections import defaultdict
+from numbers import Integral
+from uuid import uuid4
+
+import six
+
+
+@six.add_metaclass(ABCMeta)
+class _TrainingLog(object):
+    """Base class for training log.
+
+    A training log stores the training timeline, statistics and other
+    auxiliary information. Training logs can use different backends e.g.
+    in-memory Python objects or an SQLite database.
+
+    Information is stored similar to a nested dictionary, so use
+    ``log[time][key]`` to read data. An entry without stored data will
+    return an empty dictionary-like object that can be written to,
+    ``log[time][key] = value``.
+
+    Depending on the backend, ``log[time] = {'key': 'value'}`` could fail.
+    Use ``log[time].update({'key': 'value'})`` for compatibility across
+    backends.
+
+    In addition to the set of records displaying training dynamics, a
+    training log has a :attr:`status` attribute, which is a dictionary with
+    data that is not bound to a particular time.
+
+    .. warning::
+
+       Changes to mutable objects might not be reflected in the log,
+       depending on the backend. So don't use
+       ``log.status['key'].append(...)``, use ``log.status['key'] = ...``
+       instead.
+
+    Parameters
+    ----------
+    uuid : :class:`uuid.UUID`, optional
+        The UUID of this log. For persistent log backends, passing the UUID
+        will result in an old log being loaded. Otherwise a new, random
+        UUID will be created.
+
+    Attributes
+    ----------
+    status : dict
+        A dictionary with data representing the current state of training.
+        By default it contains ``iterations_done``, ``epochs_done`` and
+        ``_epoch_ends`` (a list of time stamps when epochs ended).
+
+    """
+    def __init__(self, uuid=None):
+        if uuid is None:
+            self.uuid = uuid4()
+        else:
+            self.uuid = uuid
+        if uuid is None:
+            self.status.update({
+                'iterations_done': 0,
+                'epochs_done': 0,
+                '_epoch_ends': [],
+                'resumed_from': None
+            })
+
+    @property
+    def b_uuid(self):
```

Comment:
The problem is that we need to store `resumed_from` in the status, and SQLite3 will only take it as a buffer or memoryview. The alternative is for the UUID to be a different object depending on the backend i.e. for `SQLiteLog` we would return a buffer for `log.uuid` while for `TrainingLog` we return an UUID. I'm not a big fan of that either though.